# データサイエンス100本ノック（構造化データ加工編） - Python
# for AzureNotebook

## 【注意】オリジナル版との変更点
1. Azure Notebook ではDockerを使えないので、2020.06.18時点の 100knocks-preprocess/docker/work/data にあるCSVファイルをPostgreSQLから入手する代わりに使います。
2. オリジナルのCSVデータ, geocode.csvの'latitude'列名の初めにスペースが入っていたため、それを削除しました。

   オリジナル(100knocks-preprocess ver.1.0): ' latitude' --> 'latitude'
   

4. オリジナルの解答を元に、必要のないlibraryはimportせず、そして必要なlibraryをAzureNotebookでインストールするように最初のセルを改変してあります。
5. また、SQLではなく上記CSVからデータを読み込むように'はじめに'の最初のセルを改変してあります。

## 【Azure Notebook】動かす前に
1. 本100本ノックの解答例スクリプトはPython3.6で動作検証し、それを元にLibraryの選定を行っております。メニュ-->Kernel-->Change Kernelから、Python 3.6を選んでください。Python3ではLibraryをインストールする際にエラーが発生します。

## はじめに
- 初めに以下のセルを実行してください
- 必要なライブラリのインポートと~~データベース（PostgreSQL）~~ 100knocks-preprocess/docker/work/data にあるCSVファイルからのデータ読み込みを行います。geocode.csvに変更を加えたため、またgit cloneをするとAzure Notebooksの簡便さを損なうため、筆者(noguhiro2002)のgithubレポジトリから直接読み込みます。
- pandas等、利用が想定されるライブラリは以下セルでインポートしています
- その他利用したいライブラリがあれば適宜インストールしてください（"!pip install ライブラリ名"でインストールも可能)
- オリジナルの解答例を元に、必要なライブラリーをpipでインストールします。
- 処理は複数回に分けても構いません
- 名前、住所等はダミーデータであり、実在するものではありません

In [1]:
# pipでオリジナルの解答に必要なライブラリーをインストール
!pip install --upgrade pip
!pip install -U pandas numpy scikit-learn imbalanced-learn

# pipでオリジナルの解答に必要なライブラリーをインポート
import os
import pandas as pd
import numpy as np
from datetime import datetime, date
from dateutil.relativedelta import relativedelta
import math
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from imblearn.under_sampling import RandomUnderSampler


# データを github/noguhiro2002/100knocks-preprocess/work/data フォルダよりDataframe形式でRead
df_customer = pd.read_csv('https://raw.githubusercontent.com/The-Japan-DataScientist-Society/100knocks-preprocess/master/docker/work/data/customer.csv')
df_category = pd.read_csv('https://raw.githubusercontent.com/The-Japan-DataScientist-Society/100knocks-preprocess/master/docker/work/data/category.csv')
df_product = pd.read_csv('https://raw.githubusercontent.com/The-Japan-DataScientist-Society/100knocks-preprocess/master/docker/work/data/product.csv')
df_receipt = pd.read_csv('https://raw.githubusercontent.com/The-Japan-DataScientist-Society/100knocks-preprocess/master/docker/work/data/receipt.csv')
df_store = pd.read_csv('https://raw.githubusercontent.com/The-Japan-DataScientist-Society/100knocks-preprocess/master/docker/work/data/store.csv')
df_geocode = pd.read_csv('https://raw.githubusercontent.com/noguhiro2002/100knocks-preprocess_ForColab-AzureNotebook/master/data/geocode.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


# 演習問題

---
> P-001: レシート明細のデータフレーム（df_receipt）から全項目の先頭10件を表示し、どのようなデータを保有しているか目視で確認せよ。

In [2]:
df_receipt_head = df_receipt.head(10)
print(df_receipt_head)

   sales_ymd  sales_epoch store_cd  ...  product_cd  quantity amount
0   20181103   1541203200   S14006  ...  P070305012         1    158
1   20181118   1542499200   S13008  ...  P070701017         1     81
2   20170712   1499817600   S14028  ...  P060101005         1    170
3   20190205   1549324800   S14042  ...  P050301001         1     25
4   20180821   1534809600   S14025  ...  P060102007         1     90
5   20190605   1559692800   S13003  ...  P050102002         1    138
6   20181205   1543968000   S14024  ...  P080101005         1     30
7   20190922   1569110400   S14040  ...  P070501004         1    128
8   20170504   1493856000   S13020  ...  P071302010         1    770
9   20191010   1570665600   S14027  ...  P071101003         1    680

[10 rows x 9 columns]


---
> P-002: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、10件表示させよ。

In [3]:
print(df_receipt[['sales_ymd','customer_id', 'product_cd','amount']].head(10))


   sales_ymd     customer_id  product_cd  amount
0   20181103  CS006214000001  P070305012     158
1   20181118  CS008415000097  P070701017      81
2   20170712  CS028414000014  P060101005     170
3   20190205  ZZ000000000000  P050301001      25
4   20180821  CS025415000050  P060102007      90
5   20190605  CS003515000195  P050102002     138
6   20181205  CS024514000042  P080101005      30
7   20190922  CS040415000178  P070501004     128
8   20170504  ZZ000000000000  P071302010     770
9   20191010  CS027514000015  P071101003     680


---
> P-003: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、10件表示させよ。ただし、sales_ymdはsales_dateに項目名を変更しながら抽出すること。

In [4]:
df_receipt_rename = df_receipt.rename(columns={'sales_ymd': 'sales_date'})
print(df_receipt_rename[['sales_date','customer_id','product_cd','amount']].head(10))

   sales_date     customer_id  product_cd  amount
0    20181103  CS006214000001  P070305012     158
1    20181118  CS008415000097  P070701017      81
2    20170712  CS028414000014  P060101005     170
3    20190205  ZZ000000000000  P050301001      25
4    20180821  CS025415000050  P060102007      90
5    20190605  CS003515000195  P050102002     138
6    20181205  CS024514000042  P080101005      30
7    20190922  CS040415000178  P070501004     128
8    20170504  ZZ000000000000  P071302010     770
9    20191010  CS027514000015  P071101003     680


---
> P-004: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"

In [5]:
CS = df_receipt[df_receipt['customer_id'].isin(['CS018205000001'])]
print(CS[['sales_ymd', 'customer_id','product_cd', 'amount']])

       sales_ymd     customer_id  product_cd  amount
36      20180911  CS018205000001  P071401012    2200
9843    20180414  CS018205000001  P060104007     600
21110   20170614  CS018205000001  P050206001     990
27673   20170614  CS018205000001  P060702015     108
27840   20190216  CS018205000001  P071005024     102
28757   20180414  CS018205000001  P071101002     278
39256   20190226  CS018205000001  P070902035     168
58121   20190924  CS018205000001  P060805001     495
68117   20190226  CS018205000001  P071401020    2200
72254   20180911  CS018205000001  P071401005    1100
88508   20190216  CS018205000001  P040101002     218
91525   20190924  CS018205000001  P091503001     280


---
> P-005: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 売上金額（amount）が1,000以上

In [6]:
!pip list

Package                       Version
----------------------------- --------------
absl-py                       0.12.0
alabaster                     0.7.12
albumentations                0.1.12
altair                        4.2.0
appdirs                       1.4.4
argcomplete                   2.0.0
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
arviz                         0.11.4
astor                         0.8.1
astropy                       4.3.1
astunparse                    1.6.3
atari-py                      0.2.9
atomicwrites                  1.4.0
attrs                         21.4.0
audioread                     2.1.9
autograd                      1.3
Babel                         2.9.1
backcall                      0.2.0
beautifulsoup4                4.6.3
bleach                        4.1.0
blis                          0.4.1
bokeh                         2.3.3
Bottleneck                    1.3.2
branca                        0.4.2
bs4         

In [7]:
!pip install numexpr --upgrade

In [8]:
!pip install pandas --upgrade

In [9]:
amount_1000_over = df_receipt.query("customer_id == 'CS018205000001' & amount >= 1000")
# amount_1000_over = df_receipt[(df_receipt["customer_id"]=='CS018205000001') &(df_receipt["amount"]>=1000)]
print(amount_1000_over)

       sales_ymd  sales_epoch store_cd  ...  product_cd  quantity amount
36      20180911   1536624000   S13018  ...  P071401012         1   2200
68117   20190226   1551139200   S13018  ...  P071401020         1   2200
72254   20180911   1536624000   S13018  ...  P071401005         1   1100

[3 rows x 9 columns]


---
> P-006: レシート明細データフレーム「df_receipt」から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上数量（quantity）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 売上金額（amount）が1,000以上または売上数量（quantity）が5以上

In [10]:
df_006 = df_receipt.query("customer_id == 'CS018205000001' & (amount >= 1000 | quantity >= 5)")
print(df_006[['sales_ymd', 'customer_id', 'product_cd', 'quantity', 'amount']])

       sales_ymd     customer_id  product_cd  quantity  amount
36      20180911  CS018205000001  P071401012         1    2200
9843    20180414  CS018205000001  P060104007         6     600
21110   20170614  CS018205000001  P050206001         5     990
68117   20190226  CS018205000001  P071401020         1    2200
72254   20180911  CS018205000001  P071401005         1    1100


---
> P-007: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 売上金額（amount）が1,000以上2,000以下

In [11]:
df_007 = df_receipt.query("customer_id == 'CS018205000001' & (amount >= 1000 & amount <= 2000)")
df_007[['sales_ymd', 'customer_id', 'product_cd', 'amount']]

sales_ymd     customer_id  product_cd  amount
72254   20180911  CS018205000001  P071401005    1100

---
> P-008: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 商品コード（product_cd）が"P071401019"以外

In [12]:
df_008 = df_receipt.query("customer_id == 'CS018205000001' & product_cd != 'P071401019'")
df_008[['sales_ymd', 'customer_id', 'product_cd', 'amount']]
# df_receipt.query("amount > 1")
# df_receipt[df_receipt['amount']]

sales_ymd     customer_id  product_cd  amount
36      20180911  CS018205000001  P071401012    2200
9843    20180414  CS018205000001  P060104007     600
21110   20170614  CS018205000001  P050206001     990
27673   20170614  CS018205000001  P060702015     108
27840   20190216  CS018205000001  P071005024     102
28757   20180414  CS018205000001  P071101002     278
39256   20190226  CS018205000001  P070902035     168
58121   20190924  CS018205000001  P060805001     495
68117   20190226  CS018205000001  P071401020    2200
72254   20180911  CS018205000001  P071401005    1100
88508   20190216  CS018205000001  P040101002     218
91525   20190924  CS018205000001  P091503001     280

---
> P-009: 以下の処理において、出力結果を変えずにORをANDに書き換えよ。

`df_store.query('not(prefecture_cd == "13" | floor_area > 900)')`

In [13]:
df_store.query('not(prefecture_cd == "13" | floor_area > 900)')
# df_store.query('prefecture_cd != "13" & floor_area <= 900')

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0
23   S13008        成城店             13  ...  139.5966  35.63614      883.0
24   S13020      十条仲原店             13  ...  139.7186  35.76686      801.0
25   S13043       南品川店             13  ...  139.7436  35.60981      845.0
29   S13041       八王子店             13  ...  139.4235  35.63787      810.0
30   S13017      高円寺南店             13  ...  139.6513  35.70273      841.0
38   S12013       習志野店             12  ...  140.0220  35.66122      808.0
50   S13037     南砂一丁目店             13  ...  139.8215  35.67898      814.0

[9 rows x 10 columns]

---
> P-010: 店舗データフレーム（df_store）から、店舗コード（store_cd）が"S14"で始まるものだけ全項目抽出し、10件だけ表示せよ。

In [14]:
df_store.query('store_cd.str.startswith("S14")', engine='python').head(10)

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
2    S14010        菊名店             14  ...  139.6326  35.50049     1732.0
3    S14033       阿久和店             14  ...  139.4961  35.45918     1495.0
4    S14036     相模原中央店             14  ...  139.3716  35.57327     1679.0
7    S14040       長津田店             14  ...  139.4994  35.52398     1548.0
9    S14050      阿久和西店             14  ...  139.4961  35.45918     1830.0
12   S14028       二ツ橋店             14  ...  139.4963  35.46304     1574.0
16   S14012      本牧和田店             14  ...  139.6582  35.42156     1341.0
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
19   S14022        逗子店             14  ...  139.5789  35.29642     1838.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0

[10 rows x 10 columns]

---
> P-011: 顧客データフレーム（df_customer）から顧客ID（customer_id）の末尾が1のものだけ全項目抽出し、10件だけ表示せよ。

In [15]:
df_customer.query('customer_id.str.endswith("1")', engine='python').head(10)

customer_id customer_name  ...  application_date     status_cd
1    CS037613000071         六角 雅彦  ...          20150414  0-00000000-0
3    CS028811000001        堀井 かおり  ...          20160115  0-00000000-0
14   CS040412000191         川井 郁恵  ...          20151101  1-20091025-4
31   CS028314000011        小菅 あおい  ...          20151123  1-20080426-5
56   CS039212000051        藤島 恵梨香  ...          20171121  1-20100215-4
59   CS015412000111         松居 奈月  ...          20150629  0-00000000-0
63   CS004702000041          野島 洋  ...          20170218  0-00000000-0
74   CS041515000001         栗田 千夏  ...          20160422  E-20100803-F
85   CS029313000221        北条 ひかり  ...          20180810  0-00000000-0
102  CS034312000071         望月 奈央  ...          20160106  0-00000000-0

[10 rows x 11 columns]

---
> P-012: 店舗データフレーム（df_store）から横浜市の店舗だけ全項目表示せよ。

In [16]:
df_store.query('address.str.contains("横浜")', engine='python')

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
2    S14010        菊名店             14  ...  139.6326  35.50049     1732.0
3    S14033       阿久和店             14  ...  139.4961  35.45918     1495.0
7    S14040       長津田店             14  ...  139.4994  35.52398     1548.0
9    S14050      阿久和西店             14  ...  139.4961  35.45918     1830.0
12   S14028       二ツ橋店             14  ...  139.4963  35.46304     1574.0
16   S14012      本牧和田店             14  ...  139.6582  35.42156     1341.0
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0
26   S14048      中川中央店             14  ...  139.5758  35.54912     1657.0
40   S14042       新山下店             14  ...  139.6593  35.43894     1044.0
52   S14006       葛が谷店             14  ...  139.5633  35.53573     1886.0

[11 rows x 10 columns]

---
> P-013: 顧客データフレーム（df_customer）から、ステータスコード（status_cd）の先頭がアルファベットのA〜Fで始まるデータを全項目抽出し、10件だけ表示せよ。

In [17]:
df_customer.query('status_cd.str.contains("^[A-F]")', engine='python').head(10)

customer_id customer_name  ...  application_date     status_cd
2   CS031415000172       宇多田 貴美子  ...          20150529  D-20100325-C
6   CS015414000103         奥野 陽子  ...          20150722  B-20100609-B
12  CS011215000048         芦田 沙耶  ...          20150228  C-20100421-9
15  CS029415000023         梅田 里穂  ...          20150610  D-20100918-E
21  CS035415000029         寺沢 真希  ...          20141220  F-20101029-F
32  CS031415000106        宇野 由美子  ...          20150201  F-20100511-E
33  CS029215000025         石倉 美帆  ...          20150708  B-20100820-C
40  CS033605000005         猪股 雄太  ...          20150425  F-20100917-E
44  CS033415000229        板垣 菜々美  ...          20150712  F-20100326-E
53  CS008415000145         黒谷 麻緒  ...          20150829  F-20100622-F

[10 rows x 11 columns]

---
> P-014: 顧客データフレーム（df_customer）から、ステータスコード（status_cd）の末尾が数字の1〜9で終わるデータを全項目抽出し、10件だけ表示せよ。

In [18]:
df_customer.query('status_cd.str.contains("[1-9]$")', engine='python').head(10)

customer_id customer_name  ...  application_date     status_cd
4   CS001215000145         田崎 美紀  ...          20170605  6-20090929-2
9   CS033513000180          安斎 遥  ...          20150728  6-20080506-5
12  CS011215000048         芦田 沙耶  ...          20150228  C-20100421-9
14  CS040412000191         川井 郁恵  ...          20151101  1-20091025-4
16  CS009315000023         皆川 文世  ...          20150319  5-20080322-1
22  CS015315000033        福士 璃奈子  ...          20141024  4-20080219-3
23  CS023513000066         神戸 そら  ...          20150915  5-20100524-9
24  CS035513000134         市川 美帆  ...          20150227  8-20100711-9
27  CS001515000263         高松 夏空  ...          20160812  1-20100804-1
28  CS040314000027       鶴田 きみまろ  ...          20150122  2-20080426-4

[10 rows x 11 columns]

---
> P-015: 顧客データフレーム（df_customer）から、ステータスコード（status_cd）の先頭がアルファベットのA〜Fで始まり、末尾が数字の1〜9で終わるデータを全項目抽出し、10件だけ表示せよ。

In [19]:
df_customer.query('status_cd.str.contains("^[A-F]") & status_cd.str.contains("[1-9]$")', engine='python').head(10)

customer_id customer_name  ...  application_date     status_cd
12   CS011215000048         芦田 沙耶  ...          20150228  C-20100421-9
68   CS022513000105        島村 貴美子  ...          20150320  A-20091115-7
71   CS001515000096         水野 陽子  ...          20150614  A-20100724-7
122  CS013615000053         西脇 季衣  ...          20150128  B-20100329-6
144  CS020412000161          小宮 薫  ...          20150822  B-20081021-3
178  CS001215000097        竹中 あさみ  ...          20170315  A-20100211-2
252  CS035212000007        内村 恵梨香  ...          20151013  B-20101018-6
259  CS002515000386         野田 コウ  ...          20160410  C-20100127-8
293  CS001615000372        稲垣 寿々花  ...          20170403  A-20100104-1
297  CS032512000121         松井 知世  ...          20150727  A-20100103-5

[10 rows x 11 columns]

---
> P-016: 店舗データフレーム（df_store）から、電話番号（tel_no）が3桁-3桁-4桁のデータを全項目表示せよ。

In [20]:
df_store.query('tel_no.str.contains("[0-9]{3}-[0-9]{3}-[0-9]{4}")', engine='python')

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
0    S12014       千草台店             12  ...  140.1180  35.63559     1698.0
1    S13002       国分寺店             13  ...  139.4802  35.70566     1735.0
2    S14010        菊名店             14  ...  139.6326  35.50049     1732.0
3    S14033       阿久和店             14  ...  139.4961  35.45918     1495.0
4    S14036     相模原中央店             14  ...  139.3716  35.57327     1679.0
7    S14040       長津田店             14  ...  139.4994  35.52398     1548.0
9    S14050      阿久和西店             14  ...  139.4961  35.45918     1830.0
11   S13052        森野店             13  ...  139.4383  35.55293     1087.0
12   S14028       二ツ橋店             14  ...  139.4963  35.46304     1574.0
16   S14012      本牧和田店             14  ...  139.6582  35.42156     1341.0
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
19   S14022        逗子店             14  ...  139.5789  35.29642     1838.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0
21   S13016       小金井店             13  ...  139.5094  35.70018     1399.0
22   S14034      川崎野川店             14  ...  139.5998  35.57693     1318.0
26   S14048      中川中央店             14  ...  139.5758  35.54912     1657.0
27   S12007        佐倉店             12  ...  140.1452  35.71872     1895.0
28   S14026     辻堂西海岸店             14  ...  139.4466  35.32464     1732.0
29   S13041       八王子店             13  ...  139.4235  35.63787      810.0
31   S14049      川崎大師店             14  ...  139.7327  35.53759      962.0
32   S14023        川崎店             14  ...  139.7028  35.53599     1804.0
33   S13018        清瀬店             13  ...  139.5178  35.76885     1220.0
35   S14027       南藤沢店             14  ...  139.4896  35.33762     1521.0
36   S14021       伊勢原店             14  ...  139.3129  35.40169      962.0
37   S14047       相模原店             14  ...  139.3748  35.55959     1047.0
38   S12013       習志野店             12  ...  140.0220  35.66122      808.0
40   S14042       新山下店             14  ...  139.6593  35.43894     1044.0
42   S12030        八幡店             12  ...  139.9240  35.72318     1162.0
44   S14025        大和店             14  ...  139.4680  35.43414     1011.0
45   S14045        厚木店             14  ...  139.3651  35.44182      980.0
47   S12029        東野店             12  ...  139.8968  35.65086     1101.0
49   S12053        高洲店             12  ...  139.9176  35.63755     1555.0
51   S14024        三田店             14  ...  139.5424  35.60770      972.0
52   S14006       葛が谷店             14  ...  139.5633  35.53573     1886.0

[34 rows x 10 columns]

---
> P-17: 顧客データフレーム（df_customer）を生年月日（birth_day）で高齢順にソートし、先頭10件を全項目表示せよ。

In [21]:
df_customer.sort_values('birth_day', ascending=True).head(10)

customer_id customer_name  ...  application_date     status_cd
18817  CS003813000014        村山 菜々美  ...          20160214  0-00000000-0
12328  CS026813000004         吉村 朝陽  ...          20150723  0-00000000-0
15682  CS018811000003         熊沢 美里  ...          20150403  0-00000000-0
15302  CS027803000004         内村 拓郎  ...          20151227  0-00000000-0
1681   CS013801000003         天野 拓郎  ...          20160120  0-00000000-0
7511   CS001814000022         鶴田 里穂  ...          20161012  A-20090415-7
2378   CS016815000002         山元 美紀  ...          20150629  C-20090923-C
4680   CS009815000003         中田 里穂  ...          20150421  D-20091021-E
16070  CS005813000015        金谷 恵梨香  ...          20150506  0-00000000-0
6305   CS012813000013         宇野 南朋  ...          20150712  0-00000000-0

[10 rows x 11 columns]

---
> P-18: 顧客データフレーム（df_customer）を生年月日（birth_day）で若い順にソートし、先頭10件を全項目表示せよ。

In [22]:
df_customer.sort_values('birth_day', ascending=False).head(10)

customer_id customer_name  ...  application_date     status_cd
15639  CS035114000004         大村 美里  ...          20150619  6-20091205-6
7468   CS022103000002        福山 はじめ  ...          20160909  0-00000000-0
10745  CS002113000009        柴田 真悠子  ...          20160304  0-00000000-0
19811  CS004115000014         松井 京子  ...          20161120  1-20081231-1
7039   CS002114000010          山内 遥  ...          20160920  6-20100510-1
3670   CS025115000002         小柳 夏希  ...          20160116  D-20100913-D
12493  CS002113000025        広末 まなみ  ...          20171030  0-00000000-0
15977  CS033112000003         長野 美紀  ...          20150606  0-00000000-0
5716   CS007115000006          福岡 瞬  ...          20151118  F-20101016-F
15097  CS014113000008         矢口 莉緒  ...          20150622  3-20091108-6

[10 rows x 11 columns]

---
> P-19: レシート明細データフレーム（df_receipt）に対し、1件あたりの売上金額（amount）が高い順にランクを付与し、先頭10件を抽出せよ。項目は顧客ID（customer_id）、売上金額（amount）、付与したランクを表示させること。なお、売上金額（amount）が等しい場合は同一順位を付与するものとする。

In [23]:
# df_19 = df_receipt.sort_values('amount',ascending=False)
# print(df_19[['customer_id', 'amount']].rank(method='dense').head(10))

df_19 = df_receipt

# df_19['rank'] = df_19['amount'].rank(method='max')
# df_19.sort_values('rank', ascending=False)


df_19['rank']=df_19['amount'].rank(ascending=False,method='max')

df_19.sort_values('rank',ascending=True).head(10)
# sample["rank"]=sample.groupby("ID")["セッション"].rank(ascending=True,method="first")

sales_ymd  sales_epoch store_cd  ...  quantity  amount  rank
1202    20190903   1567468800   S14011  ...         1   10925   1.0
62317   20190304   1551657600   S14046  ...         1    6800   2.0
54095   20170921   1505952000   S14028  ...         1    5780   3.0
4632    20170721   1500595200   S13015  ...         1    5480   6.0
72747   20190724   1563926400   S13009  ...         1    5480   6.0
10320   20190103   1546473600   S14021  ...         1    5480   6.0
97294   20190209   1549670400   S14021  ...         1    5440   8.0
28304   20180621   1529539200   S13004  ...         1    5440   8.0
92246   20180212   1518393600   S13009  ...         1    5280  39.0
68553   20181021   1540080000   S14040  ...         1    5280  39.0

[10 rows x 10 columns]

---
> P-020: レシート明細データフレーム（df_receipt）に対し、1件あたりの売上金額（amount）が高い順にランクを付与し、先頭10件を抽出せよ。項目は顧客ID（customer_id）、売上金額（amount）、付与したランクを表示させること。なお、売上金額（amount）が等しい場合でも別順位を付与すること。

In [24]:
df_20 = df_receipt

# df_19['rank'] = df_19['amount'].rank(method='max')
# df_19.sort_values('rank', ascending=False)


df_20['rank']=df_20['amount'].rank(ascending=False,method='first')

df_20.sort_values('rank',ascending=True).head(10)


sales_ymd  sales_epoch store_cd  ...  quantity  amount  rank
1202    20190903   1567468800   S14011  ...         1   10925   1.0
62317   20190304   1551657600   S14046  ...         1    6800   2.0
54095   20170921   1505952000   S14028  ...         1    5780   3.0
4632    20170721   1500595200   S13015  ...         1    5480   4.0
10320   20190103   1546473600   S14021  ...         1    5480   5.0
72747   20190724   1563926400   S13009  ...         1    5480   6.0
28304   20180621   1529539200   S13004  ...         1    5440   7.0
97294   20190209   1549670400   S14021  ...         1    5440   8.0
596     20181219   1545177600   S13015  ...         1    5280   9.0
11275   20180603   1527984000   S13017  ...         1    5280  10.0

[10 rows x 10 columns]

---
> P-021: レシート明細データフレーム（df_receipt）に対し、件数をカウントせよ。

In [25]:
df_receipt.shape

(104681, 10)

---
> P-022: レシート明細データフレーム（df_receipt）の顧客ID（customer_id）に対し、ユニーク件数をカウントせよ。

In [26]:
print(len(df_receipt['customer_id'].unique()))

8307


---
> P-023: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）と売上数量（quantity）を合計せよ。

In [33]:
!pip install pandas-profiling

In [34]:
!pip install --upgrade pandas_profiling

In [35]:
import pandas_profiling as pdp

ImportError: ignored

In [36]:
df_023 = df_receipt
df_023['P023'] = df_023['quantity']*df_023['amount']
df_023


# pdp.ProfileReport(df_023)

sales_ymd  sales_epoch store_cd  ...  amount      rank P023
0        20181103   1541203200   S14006  ...     158   55059.0  158
1        20181118   1542499200   S13008  ...      81   95294.0   81
2        20170712   1499817600   S14028  ...     170   51700.0  170
3        20190205   1549324800   S14042  ...      25  104339.0   25
4        20180821   1534809600   S14025  ...      90   89871.0   90
...           ...          ...      ...  ...     ...       ...  ...
104676   20180221   1519171200   S13043  ...      40  103987.0   40
104677   20190911   1568160000   S14047  ...     218   39599.0  218
104678   20170311   1489190400   S14040  ...     168   53996.0  168
104679   20170331   1490918400   S13002  ...     148   61293.0  148
104680   20190423   1555977600   S13016  ...     138   66061.0  138

[104681 rows x 11 columns]

In [37]:
df_023.query('quantity != 1')

sales_ymd  sales_epoch store_cd  ...  amount      rank  P023
22       20170306   1488758400   S14026  ...      80   95565.0   160
48       20170616   1497571200   S12030  ...     596   12406.0  1192
92       20190312   1552348800   S14022  ...     140   62398.0   280
126      20191017   1571270400   S13004  ...    2200    1348.0  4400
179      20170806   1501977600   S14033  ...      60  100623.0   120
...           ...          ...      ...  ...     ...       ...   ...
104410   20170523   1495497600   S14022  ...     444   15145.0  1332
104443   20181006   1538784000   S12013  ...     330   20518.0   660
104475   20170627   1498521600   S14025  ...      96   89131.0   192
104518   20170822   1503360000   S13043  ...     296   25377.0   592
104612   20170426   1493164800   S12007  ...     296   25378.0   592

[1108 rows x 11 columns]

---
> P-024: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに最も新しい売上日（sales_ymd）を求め、10件表示せよ。

In [38]:
df_024 = df_receipt.sort_values('sales_ymd', ascending=False).head(10)
df_024

sales_ymd  sales_epoch store_cd  ...  amount      rank P023
7707    20191031   1572480000   S13004  ...      88   91561.0   88
5836    20191031   1572480000   S13037  ...     102   76785.0  102
74566   20191031   1572480000   S14036  ...     118   72934.0  118
2212    20191031   1572480000   S13041  ...     160   54624.0  160
55541   20191031   1572480000   S13051  ...     880    9441.0  880
65157   20191031   1572480000   S13003  ...     158   55918.0  158
132     20191031   1572480000   S13019  ...      30  104116.0   30
85761   20191031   1572480000   S13001  ...      95   89603.0   95
56145   20191031   1572480000   S14025  ...     100   82833.0  100
74338   20191031   1572480000   S14006  ...     102   79199.0  102

[10 rows x 11 columns]

---
> P-025: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに最も古い売上日（sales_ymd）を求め、10件表示せよ。

In [39]:
df_025 = df_receipt.sort_values('sales_ymd', ascending=True).head(10)
df_025

sales_ymd  sales_epoch store_cd  ...  amount      rank P023
81090    20170101   1483228800   S13018  ...     248   34369.0  248
17829    20170101   1483228800   S13019  ...     378   17539.0  378
101324   20170101   1483228800   S13018  ...      40  103961.0   40
17167    20170101   1483228800   S14010  ...     148   57725.0  148
90866    20170101   1483228800   S13015  ...     100   84425.0  100
22789    20170101   1483228800   S14012  ...     138   63419.0  138
83558    20170101   1483228800   S13032  ...     110   74946.0  110
4278     20170101   1483228800   S13032  ...     218   38679.0  218
311      20170101   1483228800   S14034  ...      98   85067.0   98
63515    20170101   1483228800   S13020  ...     128   69421.0  128

[10 rows x 11 columns]

---
> P-026: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに最も新しい売上日（sales_ymd）と古い売上日を求め、両者が異なるデータを10件表示せよ。

In [40]:
# type(df_receipt['sales_ymd'])


df_026 = df_receipt.groupby('customer_id').agg({"sales_ymd":['max','min']}).reset_index()
df_026.columns = ["_".join(pair) for pair in df_026.columns]
# df_026
df_026.query('sales_ymd_max != sales_ymd_min').head(10)


# # df_receipt_max
# df_receipt_max_uni = df_receipt['sales_ymd'].unique().tolist()

# df_receipt_max_query = df_receipt.query(f'sales_ymd == {df_receipt_max_uni}')

# df_receipt_max_query

# df_receipt_max_query.query(f'sales_ymd == {df_receipt_max_uni}')




customer_id_  sales_ymd_max  sales_ymd_min
1   CS001114000005       20190731       20180503
2   CS001115000010       20190405       20171228
3   CS001205000004       20190625       20170914
4   CS001205000006       20190224       20180207
13  CS001214000009       20190902       20170306
14  CS001214000017       20191006       20180828
16  CS001214000048       20190929       20171109
17  CS001214000052       20190617       20180208
20  CS001215000005       20181021       20170206
21  CS001215000040       20171022       20170214

---
> P-027: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の平均を計算し、降順でTOP5を表示せよ。

In [57]:
df_receipt[['store_cd', 'amount']]

store_cd  amount
0        S14006     158
1        S13008      81
2        S14028     170
3        S14042      25
4        S14025      90
...         ...     ...
104676   S13043      40
104677   S14047     218
104678   S14040     168
104679   S13002     148
104680   S13016     138

[104681 rows x 2 columns]

In [59]:
df_receipt[['store_cd', 'amount']].groupby('store_cd').mean().sort_values('amount',ascending=False).head(5)




amount
store_cd            
S13052    402.867470
S13015    351.111960
S13003    350.915519
S14010    348.791262
S13001    348.470386

---
> P-028: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の中央値を計算し、降順でTOP5を表示せよ。

In [60]:
df_receipt[['store_cd', 'amount']].groupby('store_cd').median().sort_values('amount', ascending=False).head(5)

amount
store_cd        
S13052     190.0
S14010     188.0
S14050     185.0
S14040     180.0
S13003     180.0

---
> P-029: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに商品コード（product_cd）の最頻値を求めよ。

In [91]:
# df_receipt['product_cd']
df_receipt[['store_cd','product_cd']].groupby('store_cd').product_cd.apply(lambda x: x.mode())
 
# print(df_receipt.query('store_cd == "S13051" & product_cd == "P050102001"').shape)
# print(df_receipt.query('store_cd == "S13051" & product_cd == "P071003001"').shape)
# print(df_receipt.query('store_cd == "S13051" & product_cd == "P080804001"').shape)
# df_receipt.query('store_cd == "S12007"')['amount']

store_cd   
S12007    0    P060303001
S12013    0    P060303001
S12014    0    P060303001
S12029    0    P060303001
S12030    0    P060303001
S13001    0    P060303001
S13002    0    P060303001
S13003    0    P071401001
S13004    0    P060303001
S13005    0    P040503001
S13008    0    P060303001
S13009    0    P060303001
S13015    0    P071401001
S13016    0    P071102001
S13017    0    P060101002
S13018    0    P071401001
S13019    0    P071401001
S13020    0    P071401001
S13031    0    P060303001
S13032    0    P060303001
S13035    0    P040503001
S13037    0    P060303001
S13038    0    P060303001
S13039    0    P071401001
S13041    0    P071401001
S13043    0    P060303001
S13044    0    P060303001
S13051    0    P050102001
          1    P071003001
          2    P080804001
S13052    0    P050101001
S14006    0    P060303001
S14010    0    P060303001
S14011    0    P060101001
S14012    0    P060303001
S14021    0    P060101001
S14022    0    P060303001
S14023    0    P071401001


---
> P-030: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の標本分散を計算し、降順でTOP5を表示せよ。

In [97]:
df_receipt[['store_cd', 'amount']].groupby('store_cd').amount.var(ddof=0).sort_values(ascending=False).head(5)

store_cd
S13052    440088.701311
S14011    306314.558164
S14034    296920.081011
S13001    295431.993329
S13015    295294.361116
Name: amount, dtype: float64

---
> P-031: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の標本標準偏差を計算し、降順でTOP5を表示せよ。

In [98]:
df_receipt[['store_cd', 'amount']].groupby('store_cd').amount.std(ddof=0).sort_values(ascending=False).head(5)

store_cd
S13052    663.391816
S14011    553.456916
S14034    544.903736
S13001    543.536561
S13015    543.409938
Name: amount, dtype: float64

---
> P-032: レシート明細データフレーム（df_receipt）の売上金額（amount）について、25％刻みでパーセンタイル値を求めよ。

In [108]:
# df_receipt['amount']
# df_receipt.groupby('amount').quantile(0.25)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Dropping invalid columns in DataFrameGroupBy.quantile is deprecated. In a future version, a TypeError will be raised. Before calling .quantile, select only columns which should be valid for the function.
  


sales_ymd   sales_epoch  receipt_no  ...  quantity       rank     P023
amount                                         ...                              
10      20170950.00  1.506946e+09      1114.5  ...       1.0  104554.25     10.0
20      20170921.75  1.506017e+09      1109.5  ...       1.0  104496.75     20.0
25      20180103.00  1.514938e+09      1122.0  ...       1.0  104377.00     25.0
30      20170929.00  1.506643e+09      1112.0  ...       1.0  104170.00     30.0
35      20170806.00  1.501978e+09      1112.0  ...       1.0  104089.00     35.0
...             ...           ...         ...  ...       ...        ...      ...
5440    20183018.00  1.534572e+09      1149.5  ...       1.0       7.25   5440.0
5480    20180412.00  1.523534e+09      1137.0  ...       1.0       4.50   5480.0
5780    20170921.00  1.505952e+09      1192.0  ...       1.0       3.00   5780.0
6800    20190304.00  1.551658e+09      1152.0  ...       1.0       2.00   6800.0
10925   20190903.00  1.567469e+09      1192.0  ...       1.0       1.00  10925.0

[488 rows x 7 columns]

---
> P-033: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の平均を計算し、330以上のものを抽出せよ。

---
> P-034: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに売上金額（amount）を合計して全顧客の平均を求めよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。


---
> P-035: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに売上金額（amount）を合計して全顧客の平均を求め、平均以上に買い物をしている顧客を抽出せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。なお、データは10件だけ表示させれば良い。

---
> P-036: レシート明細データフレーム（df_receipt）と店舗データフレーム（df_store）を内部結合し、レシート明細データフレームの全項目と店舗データフレームの店舗名（store_name）を10件表示させよ。

---
> P-037: 商品データフレーム（df_product）とカテゴリデータフレーム（df_category）を内部結合し、商品データフレームの全項目とカテゴリデータフレームの小区分名（category_small_name）を10件表示させよ。

---
> P-038: 顧客データフレーム（df_customer）とレシート明細データフレーム（df_receipt）から、各顧客ごとの売上金額合計を求めよ。ただし、買い物の実績がない顧客については売上金額を0として表示させること。また、顧客は性別コード（gender_cd）が女性（1）であるものを対象とし、非会員（顧客IDが'Z'から始まるもの）は除外すること。なお、結果は10件だけ表示させれば良い。

---
> P-039: レシート明細データフレーム（df_receipt）から売上日数の多い顧客の上位20件と、売上金額合計の多い顧客の上位20件を抽出し、完全外部結合せよ。ただし、非会員（顧客IDが'Z'から始まるもの）は除外すること。

---
> P-040: 全ての店舗と全ての商品を組み合わせると何件のデータとなるか調査したい。店舗（df_store）と商品（df_product）を直積した件数を計算せよ。

---
> P-041: レシート明細データフレーム（df_receipt）の売上金額（amount）を日付（sales_ymd）ごとに集計し、前日からの売上金額増減を計算せよ。なお、計算結果は10件表示すればよい。

---
> P-042: レシート明細データフレーム（df_receipt）の売上金額（amount）を日付（sales_ymd）ごとに集計し、各日付のデータに対し、１日前、２日前、３日前のデータを結合せよ。結果は10件表示すればよい。

---
> P-043： レシート明細データフレーム（df_receipt）と顧客データフレーム（df_customer）を結合し、性別（gender）と年代（ageから計算）ごとに売上金額（amount）を合計した売上サマリデータフレーム（df_sales_summary）を作成せよ。性別は0が男性、1が女性、9が不明を表すものとする。
>
> ただし、項目構成は年代、女性の売上金額、男性の売上金額、性別不明の売上金額の4項目とすること（縦に年代、横に性別のクロス集計）。また、年代は10歳ごとの階級とすること。

---
> P-044： 前設問で作成した売上サマリデータフレーム（df_sales_summary）は性別の売上を横持ちさせたものであった。このデータフレームから性別を縦持ちさせ、年代、性別コード、売上金額の3項目に変換せよ。ただし、性別コードは男性を'00'、女性を'01'、不明を'99'とする。

---
> P-045: 顧客データフレーム（df_customer）の生年月日（birth_day）は日付型（Date）でデータを保有している。これをYYYYMMDD形式の文字列に変換し、顧客ID（customer_id）とともに抽出せよ。データは10件を抽出すれば良い。

---
> P-046: 顧客データフレーム（df_customer）の申し込み日（application_date）はYYYYMMD形式の文字列型でデータを保有している。これを日付型（dateやdatetime）に変換し、顧客ID（customer_id）とともに抽出せよ。データは10件を抽出すれば良い。

---
> P-047: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）はYYYYMMDD形式の数値型でデータを保有している。これを日付型（dateやdatetime）に変換し、レシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。データは10件を抽出すれば良い。

---
> P-048: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）は数値型のUNIX秒でデータを保有している。これを日付型（dateやdatetime）に変換し、レシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。データは10件を抽出すれば良い。

---
> P-049: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）を日付型（timestamp型）に変換し、"年"だけ取り出してレシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。データは10件を抽出すれば良い。

---
> P-050: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）を日付型（timestamp型）に変換し、"月"だけ取り出してレシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。なお、"月"は0埋め2桁で取り出すこと。データは10件を抽出すれば良い。

---
> P-051: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）を日付型（timestamp型）に変換し、"日"だけ取り出してレシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。なお、"日"は0埋め2桁で取り出すこと。データは10件を抽出すれば良い。

---
> P-052: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計の上、売上金額合計に対して2000円以下を0、2000円超を1に2値化し、顧客ID、売上金額合計とともに10件表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。

---
> P-053: 顧客データフレーム（df_customer）の郵便番号（postal_cd）に対し、東京（先頭3桁が100〜209のもの）を1、それ以外のものを0に２値化せよ。さらにレシート明細データフレーム（df_receipt）と結合し、全期間において買い物実績のある顧客数を、作成した2値ごとにカウントせよ。

---
> P-054: 顧客データデータフレーム（df_customer）の住所（address）は、埼玉県、千葉県、東京都、神奈川県のいずれかとなっている。都道府県毎にコード値を作成し、顧客ID、住所とともに抽出せよ。値は埼玉県を11、千葉県を12、東京都を13、神奈川県を14とすること。結果は10件表示させれば良い。

---
> P-055: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、その合計金額の四分位点を求めよ。その上で、顧客ごとの売上金額合計に対して以下の基準でカテゴリ値を作成し、顧客ID、売上金額と合計ともに表示せよ。カテゴリ値は上から順に1〜4とする。結果は10件表示させれば良い。
>
> - 最小値以上第一四分位未満
> - 第一四分位以上第二四分位未満
> - 第二四分位以上第三四分位未満
> - 第三四分位以上

---
> P-056: 顧客データフレーム（df_customer）の年齢（age）をもとに10歳刻みで年代を算出し、顧客ID（customer_id）、生年月日（birth_day）とともに抽出せよ。ただし、60歳以上は全て60歳代とすること。年代を表すカテゴリ名は任意とする。先頭10件を表示させればよい。

---
> P-057: 前問題の抽出結果と性別（gender）を組み合わせ、新たに性別×年代の組み合わせを表すカテゴリデータを作成せよ。組み合わせを表すカテゴリの値は任意とする。先頭10件を表示させればよい。

---
> P-058: 顧客データフレーム（df_customer）の性別コード（gender_cd）をダミー変数化し、顧客ID（customer_id）とともに抽出せよ。結果は10件表示させれば良い。

---
> P-059: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を平均0、標準偏差1に標準化して顧客ID、売上金額合計とともに表示せよ。標準化に使用する標準偏差は、不偏標準偏差と標本標準偏差のどちらでも良いものとする。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-060: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を最小値0、最大値1に正規化して顧客ID、売上金額合計とともに表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-061: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を常用対数化（底=10）して顧客ID、売上金額合計とともに表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-062: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を自然対数化(底=e）して顧客ID、売上金額合計とともに表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-063: 商品データフレーム（df_product）の単価（unit_price）と原価（unit_cost）から、各商品の利益額を算出せよ。結果は10件表示させれば良い。

---
> P-064: 商品データフレーム（df_product）の単価（unit_price）と原価（unit_cost）から、各商品の利益率の全体平均を算出せよ。
ただし、単価と原価にはNULLが存在することに注意せよ。

---
> P-065: 商品データフレーム（df_product）の各商品について、利益率が30%となる新たな単価を求めよ。ただし、1円未満は切り捨てること。そして結果を10件表示させ、利益率がおよそ30％付近であることを確認せよ。ただし、単価（unit_price）と原価（unit_cost）にはNULLが存在することに注意せよ。

---
> P-066: 商品データフレーム（df_product）の各商品について、利益率が30%となる新たな単価を求めよ。今回は、1円未満を四捨五入すること（0.5については偶数方向の丸めで良い）。そして結果を10件表示させ、利益率がおよそ30％付近であることを確認せよ。ただし、単価（unit_price）と原価（unit_cost）にはNULLが存在することに注意せよ。

---
> P-067: 商品データフレーム（df_product）の各商品について、利益率が30%となる新たな単価を求めよ。今回は、1円未満を切り上げること。そして結果を10件表示させ、利益率がおよそ30％付近であることを確認せよ。ただし、単価（unit_price）と原価（unit_cost）にはNULLが存在することに注意せよ。

---
> P-068: 商品データフレーム（df_product）の各商品について、消費税率10%の税込み金額を求めよ。 1円未満の端数は切り捨てとし、結果は10件表示すれば良い。ただし、単価（unit_price）にはNULLが存在することに注意せよ。

---
> P-069: レシート明細データフレーム（df_receipt）と商品データフレーム（df_product）を結合し、顧客毎に全商品の売上金額合計と、カテゴリ大区分（category_major_cd）が"07"（瓶詰缶詰）の売上金額合計を計算の上、両者の比率を求めよ。抽出対象はカテゴリ大区分"07"（瓶詰缶詰）の購入実績がある顧客のみとし、結果は10件表示させればよい。

---
> P-070: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からの経過日数を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。

---
> P-071: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からの経過月数を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。1ヶ月未満は切り捨てること。

---
> P-072: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からの経過年数を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い。（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。1年未満は切り捨てること。

---
> P-073: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からのエポック秒による経過時間を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。なお、時間情報は保有していないため各日付は0時0分0秒を表すものとする。

---
> P-074: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、当該週の月曜日からの経過日数を計算し、売上日、当該週の月曜日付とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値でデータを保持している点に注意）。

---
> P-075: 顧客データフレーム（df_customer）からランダムに1%のデータを抽出し、先頭から10件データを抽出せよ。

---
> P-076: 顧客データフレーム（df_customer）から性別（gender_cd）の割合に基づきランダムに10%のデータを層化抽出データし、性別ごとに件数を集計せよ。

---
> P-077: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客単位に合計し、合計した売上金額の外れ値を抽出せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。なお、ここでは外れ値を平均から3σ以上離れたものとする。結果は10件表示させれば良い。

---
> P-078: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客単位に合計し、合計した売上金額の外れ値を抽出せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。なお、ここでは外れ値を第一四分位と第三四分位の差であるIQRを用いて、「第一四分位数-1.5×IQR」よりも下回るもの、または「第三四分位数+1.5×IQR」を超えるものとする。結果は10件表示させれば良い。

---
> P-079: 商品データフレーム（df_product）の各項目に対し、欠損数を確認せよ。

---
> P-080: 商品データフレーム（df_product）のいずれかの項目に欠損が発生しているレコードを全て削除した新たなdf_product_1を作成せよ。なお、削除前後の件数を表示させ、前設問で確認した件数だけ減少していることも確認すること。

---
> P-081: 単価（unit_price）と原価（unit_cost）の欠損値について、それぞれの平均値で補完した新たなdf_product_2を作成せよ。なお、平均値について1円未満は四捨五入とし、0.5については偶数寄せでかまわない。補完実施後、各項目について欠損が生じていないことも確認すること。

---
> P-082: 単価（unit_price）と原価（unit_cost）の欠損値について、それぞれの中央値で補完した新たなdf_product_3を作成せよ。なお、中央値について1円未満は四捨五入とし、0.5については偶数寄せでかまわない。補完実施後、各項目について欠損が生じていないことも確認すること。

---
> P-083: 単価（unit_price）と原価（unit_cost）の欠損値について、各商品の小区分（category_small_cd）ごとに算出した中央値で補完した新たなdf_product_4を作成せよ。なお、中央値について1円未満は四捨五入とし、0.5については偶数寄せでかまわない。補完実施後、各項目について欠損が生じていないことも確認すること。

---
> P-084: 顧客データフレーム（df_customer）の全顧客に対し、全期間の売上金額に占める2019年売上金額の割合を計算せよ。ただし、販売実績のない場合は0として扱うこと。そして計算した割合が0超のものを抽出せよ。 結果は10件表示させれば良い。また、作成したデータにNAやNANが存在しないことを確認せよ。

---
> P-085: 顧客データフレーム（df_customer）の全顧客に対し、郵便番号（postal_cd）を用いて経度緯度変換用データフレーム（df_geocode）を紐付け、新たなdf_customer_1を作成せよ。ただし、複数紐づく場合は経度（longitude）、緯度（latitude）それぞれ平均を算出すること。


---
> P-086: 前設問で作成した緯度経度つき顧客データフレーム（df_customer_1）に対し、申込み店舗コード（application_store_cd）をキーに店舗データフレーム（df_store）と結合せよ。そして申込み店舗の緯度（latitude）・経度情報（longitude)と顧客の緯度・経度を用いて距離（km）を求め、顧客ID（customer_id）、顧客住所（address）、店舗住所（address）とともに表示せよ。計算式は簡易式で良いものとするが、その他精度の高い方式を利用したライブラリを利用してもかまわない。結果は10件表示すれば良い。

$$
緯度（ラジアン）：\phi \\
経度（ラジアン）：\lambda \\
距離L = 6371 * arccos(sin \phi_1 * sin \phi_2
+ cos \phi_1 * cos \phi_2 * cos(\lambda_1 − \lambda_2))
$$

---
> P-087:  顧客データフレーム（df_customer）では、異なる店舗での申込みなどにより同一顧客が複数登録されている。名前（customer_name）と郵便番号（postal_cd）が同じ顧客は同一顧客とみなし、1顧客1レコードとなるように名寄せした名寄顧客データフレーム（df_customer_u）を作成せよ。ただし、同一顧客に対しては売上金額合計が最も高いものを残すものとし、売上金額合計が同一もしくは売上実績の無い顧客については顧客ID（customer_id）の番号が小さいものを残すこととする。

---
> P-088: 前設問で作成したデータを元に、顧客データフレームに統合名寄IDを付与したデータフレーム（df_customer_n）を作成せよ。ただし、統合名寄IDは以下の仕様で付与するものとする。
>
> - 重複していない顧客：顧客ID（customer_id）を設定
> - 重複している顧客：前設問で抽出したレコードの顧客IDを設定

---
> P-閑話: df_customer_1, df_customer_nは使わないので削除する。

---
> P-089: 売上実績のある顧客に対し、予測モデル構築のため学習用データとテスト用データに分割したい。それぞれ8:2の割合でランダムにデータを分割せよ。

---
> P-090: レシート明細データフレーム（df_receipt）は2017年1月1日〜2019年10月31日までのデータを有している。売上金額（amount）を月次で集計し、学習用に12ヶ月、テスト用に6ヶ月のモデル構築用データを3セット作成せよ。

---
> P-091: 顧客データフレーム（df_customer）の各顧客に対し、売上実績のある顧客数と売上実績のない顧客数が1:1となるようにアンダーサンプリングで抽出せよ。

---
> P-092: 顧客データフレーム（df_customer）では、性別に関する情報が非正規化の状態で保持されている。これを第三正規化せよ。

---
> P-093: 商品データフレーム（df_product）では各カテゴリのコード値だけを保有し、カテゴリ名は保有していない。カテゴリデータフレーム（df_category）と組み合わせて非正規化し、カテゴリ名を保有した新たな商品データフレームを作成せよ。

---
> P-094: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

---
> P-095: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ有り
> - 文字コードはCP932

---
> P-096: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ無し
> - 文字コードはUTF-8

---
> P-097: 先に作成した以下形式のファイルを読み込み、データフレームを作成せよ。また、先頭10件を表示させ、正しくとりまれていることを確認せよ。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

---
> P-098: 先に作成した以下形式のファイルを読み込み、データフレームを作成せよ。また、先頭10件を表示させ、正しくとりまれていることを確認せよ。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ無し
> - 文字コードはUTF-8

---
> P-099: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はTSV（タブ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

---
> P-100: 先に作成した以下形式のファイルを読み込み、データフレームを作成せよ。また、先頭10件を表示させ、正しくとりまれていることを確認せよ。
>
> - ファイル形式はTSV（タブ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

# これで１００本終わりです。おつかれさまでした！